lib/data/entities.js

```diff
@@ -0,0 +1,54 @@
+// Copyright 2022 ODK Central Developers
+// See the NOTICE file at the top-level directory of this distribution and at
+// https://github.com/getodk/central-backend/blob/master/NOTICE.
+// This file is part of ODK Central. It is subject to the license terms in
+// the LICENSE file found in the top-level directory of this distribution and at
+// https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+// including this file, may be copied, modified, propagated, or distributed
+// except according to the terms contained in the LICENSE file.
+
+const { traverseXml, findOne, findAll, and, root, node, hasAttr, attr } = require('../util/xml');
+const { submissionXmlToFieldStream } = require('./submission');
+const { Form } = require('../model/frames');
+const { construct } = require('../util/util');
+
+// turn Option[Array[Option[x]]] into Array[x]; used just below in getFormFields
+const getList = (x) => x.map((xs) => xs.map((y) => y.orElse({}))).orElse([]);
+
+const getEntityDef = (xml) => {
+  const modelNode = findOne(root('html'), node('head'), node('model'));
+  return traverseXml(xml, [
+    modelNode(findOne(node('meta'), and(node('entity'), hasAttr('dataset')))(attr('dataset'))),
+    modelNode(findAll(root(), and(node('bind'), hasAttr('ref')))(attr()))
+  ]).then(([entity, bindings]) => {
+    if (entity.isEmpty()) return null;
+    const dataset = entity.get();
+    const mapping = getList(bindings)
+      .filter((bind) => ('entities:ref' in bind))
+      .map((bind) => ({
+        path: bind.nodeset.replace(/^(\/data)/, ''),
```

Comment:
Based on the XForm [spec](https://getodk.github.io/xforms-spec/#primary-instance), I believe the root node won't necessarily be `<data>`: it could be anything. So here I think we can remove whatever that first element is. How about we validate it before removing it though, checking that the first element corresponds to the child of the primary instance?